### PR TITLE
NF nib-diff function

### PR DESCRIPTION
**Pull request by chrispycheng**
- [x] refactor away script code to reside under the module codebase
- [x] changes to setup.py
- [x] Comparing two MRI file headers
- [x] allowing user to specify header fields
- [x] comparing more than two MRI file headers
- [x] unittest (diff_values)
- [x] beautifying output
- [x] unittest for other functions

### CODE REVIEW

File: nibabel/cmdline/diff.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    if img1.header.get_data_dtype() != img2.header.get_data_dtype():
+        data_dtype = (img1.header.get_data_dtype(), img2.header.get_data_dtype())
+    else:
+        return "Same data type"
+
+    if img1.header.get_data_shape() != img2.header.get_data_shape():
+        data_shape = (img1.header.get_data_shape(), img2.header.get_data_shape())
+    else:
+        return "Same data shape"
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: this is how you could get a dict from the header `{k: img1.header[k] for k in img1.header.keys()}`

just make a function like `diff_dicts(d1, d2)` which would return a dict like `{key: (d1value, d2value)}`

File: cmdline/diff.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    img1 = nibabel.load(files[0])  # load first image
+    img2 = nibabel.load(files[1])  # load second image
+
+    def diff_dicts(compare1, compare2):
+        if {i: compare1.header[i] for i in compare1.header.keys()} == {i: compare2.header[i] for i in
+                                                                       compare2.header.keys()}:
+            return
+        else:
+            for i in compare1.header.keys():
```
Comment (yarikoptic) [inline cmdline/diff.py]: there might (in theory, if e.g. those compare's aren't necessarily nifti headers) be some field in compare2 which isn't in compare1.  Please account for that (I thought we had a test for that already)

File: cmdline/diff.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    assert len(files) == 2, "ATM we can work only with two files"  # TODO #3 -- make it work for any number
+
+    # load the files headers
+    # see which fields differ
+    # call proc_file from ls, with opts.header_fields set to the fields which differ between files
+
+    img1 = nibabel.load(files[0])  # load first image
+    img2 = nibabel.load(files[1])  # load second image
+
+    def diff_dicts(compare1, compare2):
```
Comment (yarikoptic) [inline cmdline/diff.py]: move this function outside of `def main` so you could test it independently from the rest of the code

Comment (yarikoptic) [reply cmdline/diff.py]: and add a docstring

File: cmdline/diff.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        if {i: compare1.header[i] for i in compare1.header.keys()} == {i: compare2.header[i] for i in
+                                                                       compare2.header.keys()}:
+            return
+        else:
+            for i in compare1.header.keys():
+                if compare1.header[i] != compare2.header[i]:
+                    print({i: (compare1.header[i], compare2.header[i])})
```
Comment (yarikoptic) [inline cmdline/diff.py]: don't just print.  return a dict with diffs -- see the little test function which we created (`def test_diff_dicts`) which will give you an example of the desired output (a dict with tuples of values)

File: cmdline/diff.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    img1 = nibabel.load(files[0])  # load first image
+    img2 = nibabel.load(files[1])  # load second image
+
+    def diff_dicts(compare1, compare2):
+        if {i: compare1.header[i] for i in compare1.header.keys()} == {i: compare2.header[i] for i in
+                                                                       compare2.header.keys()}:
+            return
+        else:
+            for i in compare1.header.keys():
+                if compare1.header[i] != compare2.header[i]:
```
Comment (yarikoptic) [inline cmdline/diff.py]: if there is a `.get` method, use that one with a default to be None, e.g. `compare1.header.get(i, None)`

also beware that some fields might be "numpy.ndarray" so in those cases you would need to do smth like `np.any(v1 != v2)`

File: cmdline/diff.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    # see which fields differ
+    # call proc_file from ls, with opts.header_fields set to the fields which differ between files
+
+    img1 = nibabel.load(files[0])  # load first image
+    img2 = nibabel.load(files[1])  # load second image
+
+    def diff_dicts(compare1, compare2):
+        if {i: compare1.header[i] for i in compare1.header.keys()} == {i: compare2.header[i] for i in
+                                                                       compare2.header.keys()}:
+            return
```
Comment (yarikoptic) [inline cmdline/diff.py]: I don't think this check is really needed.  If nothing differs, then at the end you should just return an empty dict (see our test)

File: cmdline/diff.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
+#!python
```
Comment (yarikoptic) [inline cmdline/diff.py]: it seems that you have again managed to get confused/or confuse git ;)  `cmdline` should be a submodule under nibabel module, not top level. So all the "fixes" below alike replacing `from nibabel.cmdline` to `from cmdline` are incorrect

File: cmdline/diff.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+__author__ = 'Yaroslav Halchenko & Christopher Cheng'
+__copyright__ = 'Copyright (c) 2017 NiBabel contributors'
+__license__ = 'MIT'
+
+# these fields are processed by the __get method
+header_fields = ['sizeof_hdr', 'dim_info', 'dim', 'intent_p1', 'intent_p2', 'intent_p3', 'intent_code', 'datatype',
+                 'bitpix', 'slice_start', 'pixdim', 'vox_offset', 'scl_slope', 'scl_inter', 'slice_end', 'slice_code',
+                 'xyzt_units', 'cal_max', 'cal_min', 'slice_duration', 'toffset', 'descrip', 'aux_file', 'qform_code',
+                 'sform_code', 'quatern_b', 'quatern_c', 'quatern_d', 'qoffset_x', 'qoffset_y', 'qoffset_z', 'srow_x',
+                 'srow_y', 'srow_z', 'intent_name', 'magic']
```
Comment (yarikoptic) [inline cmdline/diff.py]: why did you need to explicitly list them?

Comment (chrispycheng) [reply cmdline/diff.py]: the .header.keys() way of accessing headers you gave to me doesn't work with certain headers - I troubleshot them and there's no problem now that I've individually listed the ones that work. Is that a problem?

Comment (yarikoptic) [reply cmdline/diff.py]: "doesn't work" is not descriptive - provide summary what exactly "isn't working". If there would be a legit reason to skip, we could just skip some of those - try/except around handling

Hard-listing makes code not generic (might be a good time to revisit beginning of fluent PYTHON book). It wouldn't allow to compare other formats with headers, it other versions of nifti file format.

Comment (chrispycheng) [reply cmdline/diff.py]: Actually it's working now so forget about the list. Noted on the hard-listing part though, I will keep that in mind for the future! Gotta focus on the bigger picture, got it! Thanks Yarik

File: nibabel/tests/test_diff.py [hunk 9]
```diff
@@ -32,17 +33,58 @@
 from .test_parrec_data import BALLS, AFF_OFF
 from .test_helpers import assert_data_similar
 
+from hypothesis import given
```
Comment (yarikoptic) [inline nibabel/tests/test_diff.py]: so hypothesis will need to be listed in dev-requirements.txt

File: nibabel/tests/test_diff.py [hunk 10]
```diff
@@ -32,17 +33,58 @@
...
 
-def test_diff_dicts():
-    assert_equal(diff_dicts('key', None, None), None)
-    assert_equal(diff_dicts('dtype', int, float), {'dtype': (int, float)})
-    assert_equal(diff_dicts(1, 2, 2), None)
-    assert_equal(diff_dicts('hdr', 1, None), {'hdr': (1, None)})
-    assert_equal(diff_dicts('array', np.array([[-3.,  0.,  0.,  3.]]), np.array([[-2.,  0.,  0.,  3.]])), {'array': (array([[-3.,  0.,  0.,  3.]]), array([[-2.,  0.,  0.,  3.]]))})
-    # TODO: mixed cases
\ No newline at end of file
+@given(st.data())
```
Comment (yarikoptic) [inline nibabel/tests/test_diff.py]: how many cases this `@given` runs through I wonder? i.e. among that whooping 1.5 sec, which tests consume most and either there is some centralized way to control how many strategies hypothesis tests "by default"?

File: nibabel/cmdline/diff.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+        Option("-H", "--header-fields",
+               dest="header_fields", default='',
+               help="Header fields (comma separated) to be printed as well (if present)"),
+    ])
+
+    return p
+
+
+def diff_dicts(key, compare1, compare2):
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: I do not see anything in this function comparing actually dictionaries -- it is just a comparison of two values for a key and then spitting out a dictionary with that key.  so it could/should be just `diff_values(value1, value2)` and `diff_dicts` should be what it used to be -- get entire dictionaries, and then internally possibly use this `diff_values` to compare values (what you now do within `main()`)

File: nibabel/cmdline/diff.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+    assert len(files) == 2, "Please enter two files"
+    # TODO #3 -- make it work for any number
+
+    img1 = nib.load(files[0])
+    img2 = nib.load(files[1])
+
+    for i in img1.header.keys():
+        if diff_dicts(i, img1.header[i], img2.header[i]) is not None:
+            print(diff_dicts(i, img1.header[i], img2.header[i]))
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: this looping should be within `diff_dicts` and I thought we had it. Moreover, keep in mind that there might be some keys present in one dict but not in the other... I thought we had those cases covered already

File: nibabel/tests/test_diff.py (hunk 9 above)
Comment (chrispycheng) [reply nibabel/tests/test_diff.py]: Done!

File: nibabel/tests/test_diff.py (hunk 10 above)
Comment (chrispycheng) [reply nibabel/tests/test_diff.py]: The documentation states that the default number of cases is 100 but I'm not really buying it. Either way, it can be adjusted. See "max_examples" here for more information: https://hypothesis.readthedocs.io/en/latest/settings.html

File: nibabel/cmdline/diff.py (hunk 11 above)
Comment (chrispycheng) [reply nibabel/cmdline/diff.py]: I just kept calling it diff_dicts because it compared differences and spit out a dict :P But with the header I don't think comparisons of dicts are necessary (but I'd love to be persuaded otherwise). Otherwise noted! Will get cracking on it :-)

File: nibabel/cmdline/diff.py (hunk 12 above)
Comment (chrispycheng) [reply nibabel/cmdline/diff.py]: Yeah those cases are covered to just return {'key': (x, None)}. At least I think it should. The list of header fields is accessed with code like img1.header['sizeof_hdr'], and if that field isn't available for either img1 or img2, I thought it should return a None. Please let me know if the truth is otherwise! And I'll work on popping the loop inside, thank you!

Comment (chrispycheng) [reply nibabel/cmdline/diff.py]: Actually scratch that - I specifically left the loop out of the diff_values function because I thought the only way it could be tested in a separate file that tested specific inputs and outputs for diff_values would be to remove the accessing of the img.header fields from the program. I can bump this loop into another method but I'd otherwise be at a loss how to make both tests and the method work with the for loop inside. Could I please get your input on this?

**Comment by chrispycheng** [comment]
Bit stonewalled ATM because the new diff_values function that can accommodate several MRI files has incorporated header field access into the function itself and so can't really be tested without an actual file on hand. Would very much appreciate a suggestion on how to tackle this problem!

**Comment by yarikoptic** [comment]
Will be traveling today/tomorrow, will try to check it out and reply by Monday... But in principle, I think nibabel has some sample files for testing it you could create some. Check out other tests

File: nibabel/cmdline/diff.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        else:
+            header_fields = opts.header_fields.split(',')
+
+        for f in header_fields:
+            if diff_values(f, header_list) is not None:
+                    print(diff_values(f, header_list))
+
+
+def main():
+    """NO DAYS OFF"""
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: ha ha... duly noted ;-)

File: nibabel/cmdline/diff.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+                pass
+            if a and b != len(diffs)-1:
+                if np.any(diffs[a] == diffs[b]):
+                    del diffs[a]
+
+    if len(diffs) > 1:
+        return {key: diffs}
+
+
+def process_file(files, opts):
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: so it is not `file` but `files`, and better be a bit more spectific so why not `diff_files`?

File: nibabel/cmdline/diff.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+    file_list = []
+    header_list = []
+
+    for f in range(len(files)):
+        file_list.append(nib.load(files[f]))
+        for h in range(len(files)):
+            header_list.append(file_list[f].header)
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: try in general, where possible (so in some cases it is warranted but most often not) , to avoid the pattern
```python
for i in range(len(X)):
   x = X[i]
```
and use
```python
for x in X:
  ...
```
this helps also to avoid the bugs... to me not clear why you append here header of the same `f` all over again

Also checkout `zip`, `enumerate` and also even the ones available in `itertools` module -- they all come handy

File: nibabel/cmdline/diff.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+                if np.any(diffs[a] == diffs[b]):
+                    del diffs[a]
+
+    if len(diffs) > 1:
+        return {key: diffs}
+
+
+def process_files(files, opts):
+
+    header_list = []
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: why not to call it just "headers"?  it used to be important/suggested to encode the type into the variable name, but with python it is somewhat the opposite (think "generic programming")

File: nibabel/cmdline/diff.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        # signals "all fields"
+        if opts.header_fields == 'all':
+            # TODO: header fields might vary across file types, thus prior sensing would be needed
+            header_fields = header_list[0].keys()
+        else:
+            header_fields = opts.header_fields.split(',')
+
+        for f in header_fields:
+            if diff_files(f, header_list) is not None:
+                    print(diff_files(f, header_list))
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: you might better off again just returning a dict with those header fields where the diff was found.  This would allow for testing it independently of how you decide to output them at the end

Comment (yarikoptic) [reply nibabel/cmdline/diff.py]: and then you could rename the function from `process-files` to a more descriptive `get_headers_diff`

File: nibabel/cmdline/diff.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    if len(diffs) > 1:
+        return {key: diffs}
+
+
+def process_files(files, opts):
+
+    header_list = []
+
+    for f in files:
+        header_list.append(nib.load(f).header)
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: express yourself (3 lines above) via 1 line list comprehension to make code more concise/expressive

File: nibabel/cmdline/diff.py [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+    return p
+
+
+def diff_files(key, inputs):
+    diffs = []
+
+    for i in range(len(inputs)):
+        if i != len(inputs)-1:
+            temp_input_1 = inputs[i]
+            temp_input_2 = inputs[i+1]
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: not sure yet why we are comparing only to the next one (`i+1`) but above 4 lines can be  1 python line:
```for input1, input2 in zip(inputs, inputs[1:])```
while also shortening the variable name
also, since you are doing access to `[key]` - assign them to variable, e.g.
```value1 = input1[key]```.

File: nibabel/cmdline/diff.py [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+                else:
+                    diffs.append(temp_input_1[key])
+            elif type(temp_input_1[key]) != type(temp_input_2[key]):
+                if i != 0 and temp_input_2 != diffs:
+                    diffs.append(temp_input_1[key])
+                    diffs.append(temp_input_2[key])
+                else:
+                    diffs.append(temp_input_1[key])
+            else:
+                pass
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: when is else? isn't then they for sure different?

File: nibabel/cmdline/diff.py [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+        if i != len(inputs)-1:
+            temp_input_1 = inputs[i]
+            temp_input_2 = inputs[i+1]
+            if np.any(temp_input_1[key] != temp_input_2[key]):
+                if i != 0 and temp_input_2 != diffs:
+                    diffs.append(temp_input_1[key])
+                    diffs.append(temp_input_2[key])
+                else:
+                    diffs.append(temp_input_1[key])
+            elif type(temp_input_1[key]) != type(temp_input_2[key]):
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: may be type check should be first ?

File: nibabel/cmdline/diff.py [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+
+        Option("-H", "--header-fields",
+               dest="header_fields", default='all',
+               help="Header fields (comma separated) to be printed as well (if present)"),
+    ])
+
+    return p
+
+
+def diff_files(key, inputs):
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: adding docstrings would help us both to understand what you are trying to accomplish here... from the input also sounds really that you are diffing for a single key, not entire files (as name suggests)

File: nibabel/cmdline/diff.py [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+                if np.all(np.isnan(diffs[a])) and np.all(np.isnan(diffs[a+1])):
+                    del diffs[a]
+            except TypeError:
+                pass
+            if a and b != len(diffs)-1:
+                if np.any(diffs[a] == diffs[b]):
+                    del diffs[a]
+
+    if len(diffs) > 1:
+        return {key: diffs}
```
Comment (yarikoptic) [inline nibabel/cmdline/diff.py]: I think above seems to be over-complicated a bit ;)
pretty much at the end you just want to know if any value (no need to encode key since also known outside) is different from any other value (e.g. the first one).  So, in human language 
```python
return any(different(x, values[0]) for x in values)
```
and you need just a simple "different" helper to reliably compare two values and outside of this function compose your "dict of diffs". I do not think you want to bother returning all pairwise differences, or some other attempt to be "smart" about the difference - would be even tricky to present, or later to associate with specific files in the output.  so [KISS](https://en.wikipedia.org/wiki/KISS_principle)

File: nibabel/cmdline/diff.py (hunk 17 above)
Comment (yarikoptic) [reply nibabel/cmdline/diff.py]: and why to invoke the same function twice to do the same? assign to some variable.

Also you could try to make these 3 lines into a dict comprehension (see above comments about what the function might return instead ;))

**Comment by chrispycheng** [comment]
TBD at this point: fix new code's buggy functionality with 2 files. Should be ready for review by Yarik after that's done to push to nibabel if I'm not mistaken. May need to add more tests

**Comment by yarikoptic** [comment]
Why to merge into this master here @chrispycheng ?

**Comment by chrispycheng** [comment]
Mistake, I am undoing it

On Sep 21, 2018, at 13:29, Yaroslav Halchenko <notifications@github.com<mailto:notifications@github.com>> wrote:


Why to merge into this master here @chrispycheng<https://github.com/chrispycheng> ?
